`compatibility-calculator.js`のアルゴリズムを実装

glact1973 opened the pull request.

glact1973 committed:
`compatibility-calculator.js`のアルゴリズムを実装

num025 commented on compatibility-calculator.js:
```diff
@@ -21,6 +21,54 @@ const calculateBestMatch = function(quizSubmissions) {
+  const candidatePoint = Math.max.apply(null , candidateMatchPoint)
+  // 最も一致するcandidateのindexを探す
+  const candidateIndex = candidateMatchPoint.indexOf(candidatePoint)  
+  return candidates[candidateIndex];
```
このロジックではTODOの中の以下の条件を満たしません。
// for ties, the last person who has the highest score is returned

なぜなら、配列の中に同じ値が複数あった場合、配列の先頭に近い方のindexが返されるからです。
参考リンク:
https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf

それと、このタスクの場合、ひとつめのforループの中でmatchPointの比較を行なうとシンプルに書ける上に、上記の指摘の回避方法もあるため良いかなと思います。

glact1973 replied:
>このロジックではTODOの中の以下の条件を満たしません。
>// for ties, the last person who has the highest score is returned

完全にこの部分見落としていました・・・
最後の人を優先ということで、lastIndexOfを利用しました。

>それと、このタスクの場合、ひとつめのforループの中でmatchPointの比較を行なうとシンプルに書ける上に、上記の指摘の回避方法もあるため良いかなと思います。

確かにそうですね・・・。
とはいえ、いろいろ遅れているので、今回は上記のようにしたいと思います。

num025 reviewed (commented):
reviewさせていただきました。

num025 commented on compatibility-calculator.js:
```diff
@@ -21,6 +21,54 @@ const calculateBestMatch = function(quizSubmissions) {
+      }
+      // 質問ごとにスコア計算
+      switch (answerMatch) {
+        case 4:
```
質問の数が変わってしまうと、このロジックでは対応できなくなってしまいます。
データから質問の数を取得するか、不正解の数をカウントした方が良いです。

glact1973 replied:
質問の数を数えてswitchさせることとしました。

glact1973 committed:
指摘事項の修正

・質問数が変わる場合でもスコア計算ができるように修正
・' for ties, the last person who has the highest score is returned' に対応（indexof -> latIndexOf）
・Eslintを用いて体裁を整えた

jonatan-alama commented:
修正確認しました。lastIndexOf なら仕様通り動きます。

もう１つのやり方として、

1) for-loop の前に最高のマッチ用の変数を用意する
```
let bestCandidate;
let bestScore = 0;
```

2) `candidateMatchPoint.push(matchPoint);` の代わりに、現在のpointはbest以上かどうかを確認して、必要な場合bestCandidateとbestScoreを更新する

3) loop終わったら、bestはすでに以上の変数にあるので、探す必要がなくなります。
